### ISSUE TITLE
Doc request: docs for CompositeEffect explain AlphaBlendingMode, and negatives / pixelformats

### ISSUE THREAD

**Issue by ljw1004** [Microsoft/Win2D#97]
I've been trying to understand CompositeEffect with EffectMode.Add

As far as I can tell, it completely ignores AlphaBlendingMode of the source.

As far as I can tell, the result is notionally just a piecewise addition:
```
result = (s1.A+s1.A, s1.R+s2.R, s1.G+s2.G, s1.B+s2.B)
```

As far as I can tell, this equation is true for floating-point surfaces. But for the default B8G8R8A8UIntNormalized surface, all negative input ARGB values are treated just as "0".

This is all pretty mysterious to me! I don't know where in the pipeline those negative values are getting dropped (or maybe just clamped), and I don't know where in the docs this should be explained. Maybe there should be an overview article that explains pipelines?

```vb
    Sub TestCompositeAdd2(canvas1 As CanvasControl)
        ' What this test does:
        '    It does Composite.Add(surface1, DiscreteTransfer(surface1,{0,-0.5,1}))
        '    It does it first on a floating point surface with an input pixel set to Gray,
        '    then does the same on an 8bit byte surface with input pixel set to Gray.
        ' What I expect:
        '    It should give similar answers in both cases, Black.
        ' What I get:
        '    R32G32B32A32Float: a=2.0 r=0.0 g=0.0 b=0.0
        '    B8G8R8A8UIntNormalized: #FF808080

        Dim defaultDpi = 96.0F
        With Nothing
            Dim surface1 As New CanvasRenderTarget(canvas1, 6, 6, defaultDpi, DirectXPixelFormat.R32G32B32A32Float, CanvasAlphaMode.Ignore)
            Dim surface2 As New CanvasRenderTarget(canvas1, 6, 6, defaultDpi, DirectXPixelFormat.R32G32B32A32Float, CanvasAlphaMode.Ignore)
            Dim e1 As New DiscreteTransferEffect With {.Source = surface1, .RedTable = {0, -0.5, 1}, .GreenTable = {0, -0.5, 1}, .BlueTable = {0, -0.5, 1}}
            Dim e2 As New CompositeEffect With {.Mode = CanvasComposite.Add}
            e2.Sources.Add(surface1) : e2.Sources.Add(e1)
            Dim r = 0.5F, g = 0.5F, b = 0.5F, a = 1.0F
            Dim buf = New Byte(16 - 1) {}
            Array.Copy(BitConverter.GetBytes(r), 0, buf, 0, 4)
            Array.Copy(BitConverter.GetBytes(g), 0, buf, 4, 4)
            Array.Copy(BitConverter.GetBytes(b), 0, buf, 8, 4)
            Array.Copy(BitConverter.GetBytes(a), 0, buf, 12, 4)
            surface1.SetPixelBytes(buf, 1, 1, 1, 1)
            Using ds = surface2.CreateDrawingSession
                ds.Blend = CanvasBlend.Copy
                ds.DrawImage(e2)
            End Using
            buf = surface2.GetPixelBytes(1, 1, 1, 1)
            r = BitConverter.ToSingle(buf, 0)
            g = BitConverter.ToSingle(buf, 4)
            b = BitConverter.ToSingle(buf, 8)
            a = BitConverter.ToSingle(buf, 12)
            Debug.WriteLine($"{surface1.Description.Format}: a={a:0.0} r={r:0.0} g={g:0.0} b={b:0.0}")
        End With

        With Nothing
            Dim surface1 As New CanvasRenderTarget(canvas1, 6, 6, defaultDpi, DirectXPixelFormat.B8G8R8A8UIntNormalized, CanvasAlphaMode.Ignore)
            Dim surface2 As New CanvasRenderTarget(canvas1, 6, 6, defaultDpi, DirectXPixelFormat.B8G8R8A8UIntNormalized, CanvasAlphaMode.Ignore)
            Dim e1 As New DiscreteTransferEffect With {.Source = surface1, .RedTable = {0, -0.5, 1}, .GreenTable = {0, -0.5, 1}, .BlueTable = {0, -0.5, 1}}
            Dim e2 As New CompositeEffect With {.Mode = CanvasComposite.Add}
            e2.Sources.Add(surface1)
            e2.Sources.Add(e1)
            surface1.SetPixelColorFs({Colors.Gray}, 1, 1, 1, 1)
            Using ds = surface2.CreateDrawingSession
                ds.Blend = CanvasBlend.Copy
                ds.DrawImage(e2)
            End Using
            Dim c = surface2.GetPixelColors(1, 1, 1, 1)(0)
            Debug.WriteLine($"{surface1.Description.Format}: {c}")
        End With

        Stop
    End Sub
```

**Comment by ljw1004** [comment]
PS. Here's the rest of my test code. I ran it, copied the output into Excel, and played around with the numbers to see if I could infer the pattern.

```vb
    Sub TestCompositeAdd(canvas1 As CanvasControl)
        Dim defaultDpi = 96.0F
        Debug.WriteLine("amode1, alpha1, rgb1, amode2, alpha2, rgb2, amodeR, alphaR, rgbR")
        For Each am1 In {CanvasAlphaMode.Ignore, CanvasAlphaMode.Premultiplied}
            For Each am2 In {CanvasAlphaMode.Ignore, CanvasAlphaMode.Premultiplied}
                For Each am3 In {CanvasAlphaMode.Ignore, CanvasAlphaMode.Premultiplied}
                    Dim test1 As New CanvasRenderTarget(canvas1, 6, 6, defaultDpi, DirectXPixelFormat.R32G32B32A32Float, am1)
                    Dim test2 As New CanvasRenderTarget(canvas1, 6, 6, defaultDpi, DirectXPixelFormat.R32G32B32A32Float, am2)
                    Dim test3 As New CanvasRenderTarget(canvas1, 6, 6, defaultDpi, DirectXPixelFormat.R32G32B32A32Float, am3)
                    Dim teste As New CompositeEffect With {.Mode = CanvasComposite.Add}
                    teste.Sources.Add(test1) : teste.Sources.Add(test2)
                    For Each a1 In {0F, 0.5F, 1.0F}
                        For Each v1 In {0F, 0.5F, 1.0F}
                            For Each a2 In {0F, 0.5F, 1.0F}
                                For Each v2 In {0F, 0.5F, 1.0F}
                                    Dim c1 = ColorF.FromArgb(a1, v1, v1, v1)
                                    Dim c2 = ColorF.FromArgb(a2, v2, v2, v2)
                                    test1.SetPixelColorFs({c1}, 0, 0, 1, 1)
                                    test2.SetPixelColorFs({c2}, 0, 0, 1, 1)
                                    Using ds = test3.CreateDrawingSession()
                                        ds.Blend = CanvasBlend.Copy
                                        ds.DrawImage(teste)
                                    End Using
                                    Dim c3 = test3.GetPixelColorFs(0, 0, 1, 1)(0)
                                    Debug.WriteLine($"{am1}, {c1.A}, {c1.R}, {am2}, {c2.A}, {c2.R}, {am3}, {c3.A}, {c3.R}")
                                Next v2
                            Next a2
                        Next v1
                    Next a1
                Next am3
            Next am2
        Next am1
        Stop
    End Sub
```

**Comment by shawnhar** [comment]
New docs covering this area are included in Win2D 0.0.22 (released today):

http://microsoft.github.io/Win2D/html/EffectPrecision.htm
